### ISSUE TITLE
 comment char ('#') in $source not found

### ISSUE THREAD

**Issue by scott0000** [rsyslog/rsyslog#2428]
the bug is same as this one
http://bugzilla.adiscon.com/show_bug.cgi?id=119
this bug fixed by modify the file runtime/ctok.c
but there is no runtime/ctok* now, cause i can't reslove this problem on my computer

i found these two commits in a row, 
the file 'runtime/ctok.c' is still existed in the first commit , but disappeared in the second commit 
eb6ee672e4f649200c84c8d74bb142b749778d47
00d2ac3d712fb037411e9856e8174e9d55b5bdbc
and there was no log to describe why to delete the file, i am so confused about this

### Expected behavior
the 'garbage' file appears

### Actual behavior
there is no 'garbage' file 

### Steps to reproduce the behavior
This is a simplified version of the rule :
if $source contains '#' then -/var/log/external/garbage

### Environment
- rsyslog version: current master

**Comment by deoren** [comment]
@scott0000 I built from master (fb750bcd39b411dcba2e7d39d7da6fe03725619f) and it's working for me. I will try to respond back later with a more detailed explanation and a sample config file that can be used to reproduce the results.

In short, I used [remote_syslog2](https://github.com/papertrail/remote_syslog2) to send a test message from a flat-file on to rsyslog via 514/tcp.

1. `cd /tmp`
1. `wget https://github.com/papertrail/remote_syslog2/releases/download/v0.20/remote-syslog2_0.20_amd64.deb`
1. `dpkg -i remote-syslog2_0.20_amd64.deb`
1. `echo 'another testing string' >>  /var/log/testinput.log`
1. `remote_syslog --hostname 'this#is#a#test' -p 514 --tcp -d 127.0.0.1 -D /var/log/testinput.log`
1. `root@ubuntu-1604-virtual-machine:/tmp/rsyslog# cat /var/log/external/garbage `

Content from the file:

```
Jan 23 17:25:28 this#is#a#test notyet.log another testing string
```

Note that with systemd present the permissions on `/var/log` are managed via a [`tmpfiles.d` drop-in](https://www.freedesktop.org/software/systemd/man/tmpfiles.d.html). Depending on your distro you may already have such a drop-in present from the stock rsyslog package provided by that distro.

In the case of Ubuntu 16.04, the drop-in overrode the systemd upstream value of 0755 back to the expected default of 0775 to allow the 'syslog' group (which rsyslog runs as when privileges are dropped) write access to the `/var/log` directory path. This allows it to create directories and files as needed.

`root@ubuntu-1604-virtual-machine:/tmp# cat /usr/lib/tmpfiles.d/00rsyslog.conf `
```
# Override systemd's default tmpfiles.d/var.conf to make /var/log writable by
# the syslog group, so that rsyslog can run as user.
# See tmpfiles.d(5) for details.

# Type Path    Mode UID  GID  Age Argument
d /var/log 0775 root syslog -
```

Could this be the issue you are encountering?

Some follow-up questions:

- What distro are you testing with?
- How did you submit the test message which had '#' as a value for the hostame?
- What does your configuration look like?
- What is the output from `getfacl /var/log` ?

**Comment by deoren** [comment]
## TL;DR

* I was able to replicate successful matching of the '#' character in `$source`
* I have provided the steps to replicate my testing method (as I could have done something wrong or have completely misunderstood the report)

## Test tools

### tcpflood

Recalling that I read about a native rsyslog test tool somewhere, I went back and first found the guide:

http://www.rsyslog.com/how-to-use-tcpflood/

I then fetched the sources and built that tool. I crafted a series of test messages for my input file, but I was unable to figure out how to override the default hostname when running the tool. I looked through the example tests where raw messages had the hostname specified there, but when sending via "network" localhost was recorded even for those raw messages. I assume I was just running the tool improperly, but because I had good results with remote_syslog2, I fell back to that.

### remote_syslog2

*Download & install*

1. `cd /tmp`
1. `wget https://github.com/papertrail/remote_syslog2/releases/download/v0.20/remote-syslog2_0.20_amd64.deb`
1. `dpkg -i remote-syslog2_0.20_amd64.deb`

## rsyslog

*Build latest master (fb750bcd39b411dcba2e7d39d7da6fe03725619f)*

* Build steps
    * [Doc](https://github.com/deoren/rsyslog-examples/blob/master/building_rsyslog_from_source.md)
    * [Script](https://github.com/deoren/rsyslog-examples/blob/master/scripts/build.sh)

* [Configuration](https://github.com/deoren/rsyslog-examples/blob/master/github_issues/i2428/rsyslog.conf)

## Test

1. Follow earlier steps to download and install the remote_syslog2 package
1. Create empty test input file for remote_syslog2
    - `touch /tmp/test_input.log`
1. Run remote_syslog2 with options to override hostname to include '#' character in hostname and send to 514/tcp on the local system
    - `remote_syslog --hostname 'this#is#a#test' -p 514 --tcp -d 127.0.0.1 -D /var/log/test_input.log`
1. Send test message from another terminal
    - `echo 'another test string' >> /var/log/test_input.log`

This output was shown in the first terminal by remote_syslog2:

```
2018-01-24 07:26:53 INFO  remote_syslog.go:51 Connecting to 127.0.0.1:514 over tcp
2018-01-24 07:26:53 ERROR  remote_syslog.go:188 Cannot forward locallog.txt, it may not exist
2018-01-24 07:26:53 ERROR  remote_syslog.go:188 Cannot forward /var/log/test_input.log, it may not exist
2018-01-24 07:27:13 INFO  remote_syslog.go:198 Forwarding file: /var/log/test_input.log
```

## Test results

`root@ubuntu-devbox:/var/log# tail -n 1 /var/log/syslog /var/log/external/garbage`
```
==> /var/log/syslog <==
Jan 24 01:28:51 this#is#a#test test_input.log another test string

==> /var/log/external/garbage <==
Jan 24 01:28:51 this#is#a#test test_input.log another test string
```

**Comment by scott0000** [comment]
i get this problem on fedora27, rsyslog-8.29.0-2.fc27.x86_64
i think you might mistakenly understand my description, and here is the detailed description
===========test one=================
step:
1.` echo "if \$source contains '#' then -/var/log/external/garbage" >> /etc/rsyslog.conf`
2.`systemctl restart rsyslog`
result:
there is no file "/var/log/external/garbage"

===========test two==================
step:
1.`echo "if \$source contains 'a' then -/var/log/external/garbage" >> /etc/rsyslog.conf`
2.`systemctl restart rsyslog`
result:
the file "/var/log/external/garbage" exists

conclusion: the char '#' in the configuration doesn't effect

the another question: i found some files "runtime/ctok.c" to fix this problem in the old version, but they were removed in 00d2ac3d712fb037411e9856e8174e9d55b5bdbc. there is no log to describe why to remove the files, can any one answer my question?

`[root@localhost ~]# getfacl /var/log`
`getfacl: Removing leading '/' from absolute path names`
`# file: var/log`
`# owner: root`
`# group: root`
`user::rwx`
`group::r-x`
`other::r-x`

**Comment by deoren** [comment]
> @scott0000: i get this problem on fedora27, rsyslog-8.29.0-2.fc27.x86_64

Thanks for the extra info. I think most Red Hat based setups generally run rsyslog as root, so permissions on `/var/log` (and its subdirectories) probably isn't as much of a concern as some other distros (Debian-based for example).

Is this the configuration you're appending your rule to?

http://pkgs.fedoraproject.org/cgit/rpms/rsyslog.git/tree/rsyslog.conf

If so, that appears to be the case for Fedora (runs as root).

> @scott0000: the another question: i found some files "runtime/ctok.c" to fix this problem in the old version, but they were removed in 00d2ac3. there is no log to describe why to remove the files, can any one answer my question?

Sorry, I don't know much about the history of the project from that time frame. Perhaps @rgerhards can shed some light on the changes.

> ===========test one=================
> step:
> 1.echo "if \$source contains '#' then -/var/log/external/garbage" >> /etc/rsyslog.conf

Just to make sure we're on the same page, are you referring to the hostname of the sending system containing the '#' character and then again the 'a' character as you've shown in your examples re the`$source` [message property](http://www.rsyslog.com/doc/v8-stable/configuration/properties.html)?

Can you illustrate how you generated/submitted your test message for rsyslog to evaluate?

**Comment by rgerhards** [comment]
@deoren I think @scott0000 means that ```#``` always initiates a comment, even if inside a literal. I'll craft a quick testbench test soon to check.

**Comment by scott0000** [comment]
> Is this the configuration you're appending your rule to?
> http://pkgs.fedoraproject.org/cgit/rpms/rsyslog.git/tree/rsyslog.conf

yes, it is

my problem is the same as this http://bugzilla.adiscon.com/show_bug.cgi?id=119
you may understand it easier in the bugzilla

and see all the steps i do below
`# append “if $source contains '#' then -/var/log/external/garbage” in to rsyslog.conf`
`[root@localhost external]# vim /etc/rsyslog.conf`
`[root@localhost external]# systemctl restart rsyslog`
`[root@localhost external]# ls /var/log/external/g*`
`ls: cannot access /var/log/external/g*: No such file or directory`

`# change the last '#' to 'a' in rsyslog.conf`
`[root@localhost external]# vim /etc/rsyslog.conf`
`[root@localhost external]# systemctl restart rsyslog`
`[root@localhost external]# ls /var/log/external/g*`
`/var/log/external/garbage`

**Comment by scott0000** [comment]
@rgerhards yes, that is what i want to say

**Comment by deoren** [comment]
> my problem is the same as this http://bugzilla.adiscon.com/show_bug.cgi?id=119
> you may understand it easier in the bugzilla

> @rgerhards: @deoren I think @scott0000 means that # always initiates a comment, even if inside a literal. I'll craft a quick testbench test soon to check.

Ah! I see what is being said now. I completely misunderstood.

Regarding this behavior, I'm not seeing that with master (fb750bcd39b411dcba2e7d39d7da6fe03725619f).

@scott0000 Are you building from master (as your original report indicated) or are you using the version of rsyslog provided by fedora 27, `rsyslog-8.29.0-2.fc27.x86_64` ?

**Comment by scott0000** [comment]
@deoren i am using the version of rsyslog provided by fedora 27, rsyslog-8.29.0-2.fc27.x86_64

**Comment by rgerhards** [comment]
> Regarding this behavior, I'm not seeing that with master (fb750bc).

I can reproduce an error message with above commit. Not sure on the cause, though.

**Comment by rgerhards** [comment]
This has to do with support for obsolete legacy inside the lexer. Not yet seeing clearly what is wrong, but pretty sure it is in that area. 

@scott0000 the ctok files are long gone, they have been replaced by a proper grammar via flex/bison.

**Comment by rgerhards** [comment]
I shot myself into the foot -- my testbench test uses ```'```quote incorrectly, what essentially will result in an unquoted ```#```. With the test script fixed, all works well.

@scott0000 I now also see what is wrong: ```#``` is not permitted in hostnames, so ```$source``` will never contain ```#```. Add this line to rsyslog.conf:
```
action(type="omfile" file="/var/log/msgdebug.log" template="RSYSLOG_DebugFormat")
```
and post the lines that match one entry that you think contains your condition.

**Comment by scott0000** [comment]
@rgerhards 
ah! that resloves my problem
thanks a lot for your help!
